packages/flutter/lib/src/widgets/sliver.dart

```diff
@@ -544,22 +612,72 @@ class SliverChildListDelegate extends SliverChildDelegate {
   /// The widgets to display.
   final List<Widget> children;
 
+  /// A map to cache key to index lookup for children
+  ///
+  /// _keyToIndex[null] is used as current index during the lazy loading process
+  /// in [_findChildIndex]. _keyToIndex should never be used for looking up null key.
+  final Map<Key, int> _keyToIndex;
+
+  bool get _isConstantInstance => _keyToIndex == null;
+
+  int _findChildIndex(Key key){
+    if(_isConstantInstance) {
+      return -1;
+    }
+    // This function should only be called when any child has key during performRebuild.
+    // We do lazy loading in this function.
+    if(!_keyToIndex.containsKey(key)) {
+      int index = _keyToIndex[null];
+      while(index < children.length) {
+        final Widget child = children[index];
+        if (child.key != null) {
+          _keyToIndex[child.key] = index;
+        }
+        if (child.key == key) {
+          // Record current index for next function call.
+          _keyToIndex[null] = index + 1;
+          return index;
+        }
+        index++;
+      }
+      _keyToIndex[null] = index;
+    } else {
+      return _keyToIndex[key];
+    }
+    return -1;
+  }
+
+  @override
+  int findIndexByKey(Key key) {
+    assert(key != null);
+    Key childKey;
+    if (key is _SaltedValueKey) {
+      // in this case, child key is wrapped with _SaltedValueKey
+      final _SaltedValueKey saltedValueKey = key;
+      childKey = saltedValueKey.value;
+    } else {
+      childKey = key;
+    }
+    return _findChildIndex(childKey);
+  }
+
   @override
   Widget build(BuildContext context, int index) {
     assert(children != null);
     if (index < 0 || index >= children.length)
       return null;
     Widget child = children[index];
+    final Key key = child.key != null? _SaltedValueKey(child.key) : null;
     assert(child != null);
     if (addRepaintBoundaries)
-      child = RepaintBoundary.wrap(child, index);
+      child = RepaintBoundary(child: child, key: key);
     if (addSemanticIndexes) {
       final int semanticIndex = semanticIndexCallback(child, index);
       if (semanticIndex != null)
-        child = IndexedSemantics(index: semanticIndex + semanticIndexOffset, child: child);
+        child = IndexedSemantics(index: semanticIndex + semanticIndexOffset, child: child, key: key);
     }
     if (addAutomaticKeepAlives)
-      child = AutomaticKeepAlive(child: child);
+      child = AutomaticKeepAlive(child: child, key: key);
```

Comment:
similar comment here as above, but maybe we can improve it further by merging these two methods into one somehow and avoid the code duplication

Reply:
the only way i can think of is have a subclass of SliverChildDelegate, maybe called SliverKeepAliveChildDelegate and implement the logic in this class, and let these two delegators inherit from it. but it kinda have too many layer of abstraction. what do you think?